tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Logging/NLog/DirectSubmission/LogFactoryGetConfigurationForLoggerInstrumentation.cs

```diff
@@ -44,19 +56,105 @@ public class LogFactoryGetConfigurationForLoggerInstrumentation
+
+            _ = instance.TryDuckCast<ILogFactoryProxy>(out var logFactoryProxy);
+            _ = instance.TryDuckCast<ILogFactoryPre43Proxy>(out var logFactoryPre43Proxy);
+
+            if (logFactoryPre43Proxy is null)
+            {
+                Log.Warning("Failed to DuckCast the log factory for NLog - Agentless logging and logs injection for NLog won't be functional.");
+                return CallTargetState.GetDefault();
+            }
+
+            // when logging is being stopped/shutdown this instrumentation gets called again
+            // for later versions of NLog, the configuration passed is "null"
+            // but we have an "_isDisposing" field that is set to true in the LogFactory that we can check
+            // for older versions of NLog, the configuration passed is the _previous_ configuration used
+            // and there is no "_isDisposing" field that we can check, so using the ConfigurationTable to keep track
+
+            if (configuration is not null && ConfigurationTable.TryGetValue(configuration, out _))
+            {
+                Log.Debug("We've created a configuration already for NLog - skipping.");
+                return CallTargetState.GetDefault();
+            }
+
+            if (logFactoryProxy is not null && logFactoryProxy.IsDisposing)
+            {
+                // when logging is stopped (e.g., shutdown) the configuration will be set to null
+                // and this instrumentation will get hit again, so to avoid creating a new configuration
+                // and adding the Datadog Direct Submission target we just need to exit here.
+                Log.Debug("NLog LogFactory was being disposed so skipping the logging configuration instrumentations.");
+                return CallTargetState.GetDefault();
+            }
+
+            // we don't want to do logs injection with our custom configuration that we create as there won't be any targets
+            if (tracerManager.Settings.LogsInjectionEnabledInternal && configuration is not null)
+            {
+                LogsInjectionHelper<TTarget>.ConfigureLogsInjection(configuration);
+            }
+
+            // if there isn't a configuration AND we have DirectLogSubmission enabled, create a configuration
+            var setConfigurationRequired = false; // indicate whether we need to set LogFactory.Configuration
+            if (tracerManager.DirectLogSubmission.Settings.IsIntegrationEnabled(IntegrationId.NLog)
+             && configuration is null)
+            {
+                Log.Debug("NLog configuration was null - creating one to allow Direct Log Submission");
+                var loggingConfigurationInstance = Activator.CreateInstance(typeof(TLoggingConfiguration));
+                if (loggingConfigurationInstance is null)
+                {
+                    Log.Warning("Failed to create instance of NLog.Config.LoggingConfiguration Type");
+                    return CallTargetState.GetDefault();
```

Comment:
We've already setup logs injection at this point, so we probably don't want to exit immediately (so that we record the "integration generated span" etc). Also, we know this can't be `null` - it'll throw instead if there's an error! 😄

Reply:
That makes sense about not wanting to exit immediately.
For some reason I thought that `Activator.CreateInstance` could return `null`, but I'm not seeing that in the docs anywhere except for attempting to create some nullable type, so thanks for that

I think I should just be able to remove this? Or should I wrap it in a try/catch?

Reply:
I would probably just remove the null checks, and I would just play it safe by having a "top-level" try-catch to catch anything catastrauphic

Reply:
I shall do that